Support Peer Pod VMs

yoheiueda opened the pull request:
This PR contains necessary changes in Kata shim and agent to support Peer Pod VMs. Peer Pod VMs are sandbox VMs that are not managed by local hypervisor, but by remote hypervisor. Typically, peer pod VMs are VM instances of IaaS Cloud.

The detailed description of Peer Pod VMs will be posted at https://github.com/confidential-containers/cloud-api-adaptor. 

The main changes are as follows.

* A new hypervisor type `remote` is introduced to support a out-of-tree hypervisor driver, which is running as a external process and communicates with the shim using TTRPC via a Unix domain socket.
* A new agent socket type is also introduced to support Unix domain socket in addition to VSock.
* Recursive file copying of shared files is implemented to support Kubernetes volumes

Note that this PR also includes some work around fixes for issues we encountered to run our prototype.

A minimum kata shim configuration (`/etc/kata-containers/configuration.toml`) looks like this.
```
[runtime]
internetworking_model = "none"
disable_new_netns = true
enable_pprof = true
enable_debug = true
[hypervisor.remote]
remote_hypervisor_socket = "/run/peerpod/hypervisor.sock"
[agent.kata]
```
A kata shim and external hypervisor driver process communicates using Unix Domain Socket `/run/peerpod/hypervisor.sock`.

### Architecture
![Architecture](https://raw.githubusercontent.com/confidential-containers/cloud-api-adaptor/staging/docs/architecture.png)

fengwang666 commented:
@yoheiueda This is interesting! I would love to know more about this project. A few questions, 
1. How do you deal with the slowness of cloud VM provisioning? Would it make Kata container really slow?
2. What are the other potential feature and performance tradeoffs?
3. Is there a detail security analysis regarding this model?

yoheiueda commented:
@fengwang666

> 1. How do you deal with the slowness of cloud VM provisioning? Would it make Kata container really slow?

I did experiments using IBM Cloud VPC, and VM provisioning takes about 1 minute.  At least with this range of provisioning time, Kata containers and Kubernetes work fine without timeout errors. 

If we need faster provisioning, we need to implement some kind of optimization such as preallocation of VMs. VM preallocation are possibly not suitable with confidential computing scenarios, and we may need some ideas to implement it.

> 2. What are the other potential feature and performance tradeoffs?

We need a network tunnel between a worker node and pod VMs. All communication from/to a pod VM goes to its worker node VM. This may degrade network performance. We can implement some optimization to alleviate this network performance issue.

> Is there a detail security analysis regarding this model?

We haven't done yet, but I agree that we need security analysis.

The point here is that a pod workload does not trust worker nodes and network at all. We provide a network tunnel between worker node and pod VMs, but a pod workload must not trust the network. A worker node provides volume contents of ConfigMap and Secret, but a pod workload must not trust the integrity of the volume contents. Pod workloads must checks data integrity by themselves using security measures such as TLS and digital signatures.

fitzthum commented:
> If we need faster provisioning, we need to implement some kind of optimization such as preallocation of VMs. VM preallocation are possibly not suitable with confidential computing scenarios, and we may need some ideas to implement it.

I've thought about pre-alllocation for CC a little bit.

I think the potential for pre-allocation varies by platform. In general CC is relatively well-suited for pre-allocating VMs because the initrd is usually generic (even between different tenants). Secure Execution uses an encrypted image, which is less generic, but for other platforms you can potentially boot the VM before you know what workloads are going to run or even who will own the workloads. That said, for SEV(-ES) you would have to pause the guest before the secret injection phase, which means that you won't be able to load the kernel. You would at least get to save the time it takes the PSP to pre-encrypt the firmware (a serialized operation that can bog things down if you start a bunch of VMs at once). Of course, with peer pods a CSP would have to allow you to request VMs in this very specific manner and it's not clear how you would get secret injection to work via a CSP anyway. For TDX and SNP I think things might be easier.

arronwy commented:
Can one of the admins verify this patch?

yoheiueda commented:
@bpradipt @stevenhorsman I've rebased the PR to the latest CCv0 branch.

yoheiueda commented:
A workaround patch  https://github.com/kata-containers/kata-containers/pull/3697/commits/9a13b1adad35ca7d3a369a4eeacda35c4daf9c76 is still required for the "pause" container issue https://github.com/kata-containers/kata-containers/issues/4477 of CCv0 containerd.

stevenhorsman commented:
cc @kata-containers/architecture-committee - we are looking to get this PR for remote-hypervisor support reviewed soon and wanted to cover the topic of Peer pods in the AC meeting as part of that process

pmores commented on src/runtime/pkg/katautils/config.go:
```diff
@@ -96,6 +97,7 @@ type hypervisor struct {
 	GuestHookPath                  string   `toml:"guest_hook_path"`
 	GuestMemoryDumpPath            string   `toml:"guest_memory_dump_path"`
 	SeccompSandbox                 string   `toml:"seccompsandbox"`
+	RemoteHypervisorSocket         string   `toml:"remote_hypervisor"`
```
Wouldn't it be clearer to call the TOML key simply `remote_hypervisor_socket`?

yoheiueda replied:
I agree. I'll change it to `remote_hypervisor_socket`

yoheiueda replied:
I renamed the option name to `remote_hypervisor_socket`, and also rebased the branch to the latest `CCv0` branch.

pmores replied:
@yoheiueda Thanks!

yoheiueda commented on src/runtime/pkg/katautils/config.go:
```diff
@@ -986,6 +989,18 @@ func newClhHypervisorConfig(h hypervisor) (vc.HypervisorConfig, error) {
+		// TODO: The values below are dummy ones to bypass the checks in validateHypervisorConfig
+		KernelPath:            "dummy",
+		ConfidentialGuest:     true,
+		HypervisorMachineType: virtcontainers.QemuCCWVirtio,
```
Remote hypervisor is architecture agnostic. These values are dummy values to bypass the checks in `validateHypervisorConfig` of `virtcontainers/hypervisor_config_linux.go`.

This is just a work around, so I will introduce a flag in `HypervisorConfig` to indicate whether remote hypervisor is enable or not, and then skip the checks in `validateHypervisorConfig` when the flag is enabled.

yoheiueda replied:
I changed `validateHypervisorConfig` to check whether `RemoteHypervisorSocket` is specified or not. If it is set, `validateHypervisorConfig` skips the remaining checks.

```
func checkHypervisorConfig(config vc.HypervisorConfig) error {
	if config.RemoteHypervisorSocket != "" {
		return nil
	}
```

yoheiueda commented on src/runtime/virtcontainers/kata_agent.go:
```diff
@@ -75,7 +76,7 @@ const (
 var (
 	checkRequestTimeout          = 30 * time.Second
-	defaultRequestTimeout        = 60 * time.Second
+	defaultRequestTimeout        = 300 * time.Second
```
OK, I see. I'll change the code to use a lager timeout value only when remote hypervisor is enabled.

yoheiueda replied:
I changed timeout value only for the remote hypervisor type.

bpradipt commented on src/runtime/virtcontainers/kata_agent.go:
```diff
@@ -1262,6 +1272,31 @@ func (k *kataAgent) createContainer(ctx context.Context, sandbox *Sandbox, c *Co
+		}
+
+		//Send PullImageRequest only for "pause" image
+		if image == "k8s.gcr.io/pause:3.2" {
```
@yoheiueda for crio we'll need to pull all the images including pause. This ensures no changes to the crio side. Can we enable pullImageRequest for all the images in case of remote hypervisor ?

yoheiueda replied:
I am going to implement image pulling in the cloud-api-adaptor side. https://github.com/confidential-containers/cloud-api-adaptor/issues/88

If cri-o provides image name in annotations, cloud-api-adaptor will be able to send pull image requests for all container images including pause.

The current code to get container image name in kata shim is this one.
https://github.com/yoheiueda/kata-containers/blob/8f176024bcd7d78b81439df75824e068b31eaddd/src/runtime/virtcontainers/kata_agent.go#L1316-L1338

For some reason, containerd v1.6.6-15 does not provide annotation for image name, so we still need the custom conatinerd.

@bpradipt do you know whether cri-o provide image name as annotation `io.kubernetes.cri-o.ImageName`?

bpradipt replied:
@yoheiueda yes, the annotation is provided. 
This is from a running system
```
        "io.kubernetes.cri-o.ImageName": "quay.io/karmab/origin-coredns:latest",
        "io.kubernetes.cri-o.Name": "k8s_coredns_coredns-test-worker-0.ocp411.com_openshift-infra_dbbb729d580e7461e98ebf6dba93ab7c_1",
        "io.kubernetes.cri-o.ContainerType": "container",
        "io.kubernetes.cri-o.Stdin": "false",
```

yoheiueda replied:
I implemented https://github.com/confidential-containers/cloud-api-adaptor/issues/88, and remove the work around from kata shim.

I also added the cri-o support.

yoheiueda commented:
I rebased the code to the latest CCv0 branch.

I also added the `remote_hypervisor_timeout` parameter to specify a custom StartVM timeout value in seconds.

https://github.com/kata-containers/kata-containers/blob/6656c06924cc3b137f2086342538edaaeb851a96/src/runtime/pkg/katautils/config.go#L133

stevenhorsman commented:
/test

wainersm commented on src/runtime/cmd/kata-runtime/kata-check_s390x.go:
```diff
@@ -55,7 +55,14 @@ func kvmIsUsable() error {
+		return kvmIsUsable()
+	case "remote":
+		return nil
+	default:
```
Aren't Firecracker and Cloud Hypervisor supported on s390x?

yoheiueda replied:
They will possibly be supported on s390x. I'll fix the logic.

yoheiueda replied:
done

wainersm commented on src/runtime/pkg/katautils/config.go:
```diff
@@ -1564,6 +1578,11 @@ func checkFactoryConfig(config oci.RuntimeConfig) error {
 // config.
 func checkHypervisorConfig(config vc.HypervisorConfig) error {
+
+	if config.RemoteHypervisorSocket != "" {
```
Would make sense to have some silly checks like "timeout is unsigned int greater than zero"?

yoheiueda replied:
OK. I'll add such checks.

yoheiueda replied:
RemoteHypervisorSocket is an unsigned variable, and zero means no timoeout, so we don't need to non-negative check here.

wainersm replied:
Ok, makes sense. Thanks for addressing my comment!

wainersm commented on src/runtime/virtcontainers/kata_agent.go:
```diff
@@ -722,30 +729,38 @@ func (k *kataAgent) startSandbox(ctx context.Context, sandbox *Sandbox) error {
+		kmodules = setupKernelModules(k.kmodules)
 	}
 
 	storages := setupStorages(ctx, sandbox)
```
Sorry to ask: does the storage setup keeps the same for remote hypervisor?

yoheiueda replied:
[`setupStorage`](https://github.com/kata-containers/kata-containers/blob/6656c06924cc3b137f2086342538edaaeb851a96/src/runtime/virtcontainers/kata_agent.go#L804-L872) does nothing for remote hypervisor, since it does not support file system sharing.

I will change the code to call `setupStorage` unless remote hypervisor is used.

yoheiueda replied:
I investigated the code, and found that `setupStorage` must be called even for remote hypervisor, so I keep the old code.

The following code needs to be called to prepare shm device on pod vm.
https://github.com/kata-containers/kata-containers/blob/f30931cda97fbbf6253eac97ed2858327ae9493e/src/runtime/virtcontainers/kata_agent.go#L851-L862

wainersm commented on src/runtime/virtcontainers/pkg/agent/protocols/client/client.go:
```diff
@@ -237,6 +238,11 @@ func parse(sock string) (string, *url.URL, error) {
 		grpcAddr = HybridVSockScheme + ":" + hvsocket[0]
+	case RemoteSockScheme:
+		if addr.Host != "" {
+			return "", nil, grpcStatus.Errorf(codes.InvalidArgument, "Invalid remote sock scheme: %s", sock)
```
I suggest to explain it is invalid because the host address should be nil, if I understood correctly :)

yoheiueda replied:
OK. I'll add such explanation in the error message.

yoheiueda replied:
done

wainersm commented on src/runtime/virtcontainers/remote.go:
```diff
@@ -0,0 +1,285 @@
+	return err
+}
+
+func (rh *remoteHypervisor) PauseVM(ctx context.Context) error {
```
do you plan to implement any of the following not implemented functions still on this PR?

yoheiueda replied:
I am thinking of a small start. I have implemented a minimum set of functions to run our demo of peer pod VMs.

If some important functions are missing in my PR, I am desired to implement such functions.

snir911 commented on src/runtime/virtcontainers/remote.go:
```diff
@@ -0,0 +1,287 @@
+
+func (rh *remoteHypervisor) GetPids() []int {
+	// TODO
+	return []int{99999}
```
This pid is used by crio to fetch startTime, could you do please do something like:
```
diff --git a/src/runtime/virtcontainers/remote.go b/src/runtime/virtcontainers/remote.go
index 818da5167..ea1b3718f 100644
--- a/src/runtime/virtcontainers/remote.go
+++ b/src/runtime/virtcontainers/remote.go
@@ -8,6 +8,7 @@ import (
        "fmt"
        "net"
        "time"
+       "os"
 
        cri "github.com/containerd/containerd/pkg/cri/annotations"
        "github.com/containerd/ttrpc"
@@ -249,8 +250,8 @@ func (rh *remoteHypervisor) setConfig(config *HypervisorConfig) error {
 }
 
 func (rh *remoteHypervisor) GetPids() []int {
-       // TODO
-       return []int{99999}
+       // meanwhile let's use shim pid as it used by crio to fetch start time
+       return []int{os.Getpid()}
 }
 
 func (rh *remoteHypervisor) GetVirtioFsPid() *int {
 ```

yoheiueda replied:
I've included this change.

yoheiueda commented:
I've rebased the code to the latest CCv0 branch, and also included new changes based on review comments.

wainersm commented:
@yoheiueda thanks for addressing all my comments. As I mentioned before somewhere I am not an expert on this area, but my review didn't find any obvious problem. While it looks good to me, I prefer others with more knowledge to approve it for merging.

stevenhorsman commented:
/test

bergwolf commented on src/runtime/virtcontainers/fs_share_linux.go:
```diff
@@ -224,6 +224,52 @@ func (f *FilesystemShare) Cleanup(ctx context.Context) error {
+	return file.Readdirnames(0)
+}
+
+func (f *FilesystemShare) recursiveCopy(ctx context.Context, src, dst string) (bool, error) {
```
Is it possible to use https://pkg.go.dev/path/filepath#WalkDir instead? It should be able to save some lines.

yoheiueda replied:
done

bergwolf commented on src/runtime/virtcontainers/kata_agent.go:
```diff
@@ -2099,40 +2120,56 @@ func (k *kataAgent) setGuestDateTime(ctx context.Context, tv time.Time) error {
+	switch sflag := st.Mode & unix.S_IFMT; sflag {
+	case unix.S_IFREG:
+		var err error
+		b, err = ioutil.ReadFile(src)
```
We should not read the file in one shot but it is not this PR's fault...

yoheiueda replied:
I agree. I'll add a TODO comment here for now. We can fix the issue in another PR. 

I think we need to update the protobuf definition to support file range in CopyFileRequest.

yoheiueda replied:
done

bergwolf reviewed (commented):
Generally lgtm. Just one minor comment.

stevenhorsman commented:
/test-fedora

yoheiueda commented:
@bergwolf I've rebased the code to the latest CCv0 branch, and also included new changes based on your review comments.

bergwolf reviewed (approved):
lgtm! Thanks!

yoheiueda commented:
@bergwolf Thank you very much!

stevenhorsman commented:
/test

egernst commented:
Sorry I haven't had much time to review. I reviewed first couple of patches and didn't see any major issues. Since this is for CCv0 branch, I don't have any concerns with putting in, and I will continue to review the hypervisor interface portion next.  Thanks for driving this ya'll!

pmores reviewed (approved):
Look good to me (to the admittedly limited extent I can judge), and thanks for addressing my comment!

stevenhorsman commented:
> Sorry I haven't had much time to review. I reviewed first couple of patches and didn't see any major issues. Since this is for CCv0 branch, I don't have any concerns with putting in, and I will continue to review the hypervisor interface portion next. Thanks for driving this ya'll!

Thanks Eric, we'll get this merged in now, but as always can fix any problems in later PRs. Thanks for everyone's support!
